pkg/sink/kafka/sarama_factory.go

```diff
@@ -90,14 +105,22 @@ func (f *saramaFactory) SyncProducer(ctx context.Context) (SyncProducer, error)
 		return nil, err
 	}
 	config.MetricRegistry = f.registry
-	p, err := sarama.NewSyncProducer(f.option.BrokerEndpoints, config)
+
+	client, err := sarama.NewClient(f.option.BrokerEndpoints, config)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return &saramaSyncProducer{
+	p, err := sarama.NewSyncProducerFromClient(client)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sp := saramaSyncProducer{
 		id:       f.changefeedID,
 		producer: p,
-	}, nil
+		done:     make(chan struct{}),
+	}
+	go keepConnAlive(client, sp.done)
```

Comment:
ditto, do not spawn goroutine here.

DDLSink.WriteCheckpointTs method, is called periodically, you can call the heartbeat method here